### PR TITLE
Use best icon possible as ZIM illustration

### LINKED ISSUE #352: Retrieving upstream favicon to set illustration 48x48 is not smart enough
Scraping https://womenshistory.si.edu/ with an extensive set of good favicon/illustrations:
```
<link rel="icon" sizes="16x16" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-16x16.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-16x16.png)" />
<link rel="icon" sizes="32x32" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-32x32.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-32x32.png)" />
<link rel="icon" sizes="96x96" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-96x96.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-96x96.png)" />
<link rel="icon" sizes="192x192" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/android-chrome-192x192.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/android-chrome-192x192.png)" />
<link rel="apple-touch-icon" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-60x60.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-60x60.png)" />
<link rel="apple-touch-icon" sizes="72x72" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-72x72.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-72x72.png)" />
<link rel="apple-touch-icon" sizes="76x76" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-76x76.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-76x76.png)" />
<link rel="apple-touch-icon" sizes="114x114" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-114x114.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-114x114.png)" />
<link rel="apple-touch-icon" sizes="120x120" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-120x120.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-120x120.png)" />
<link rel="apple-touch-icon" sizes="144x144" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-144x144.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-144x144.png)" />
<link rel="apple-touch-icon" sizes="152x152" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-152x152.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-152x152.png)" />
<link rel="apple-touch-icon" sizes="180x180" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-180x180.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-180x180.png)" />
<link rel="apple-touch-icon-precomposed" sizes="180x180" href="[https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-precomposed.png](view-source:https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/apple-icon-precomposed.png)" />
```
... But warc2zim seems not able to find a good one
image

I suspect it does not looks this new favicon entries

Might be related to https://github.com/openzim/warc2zim/issues/120

### PR DESCRIPTION
**Pull request by benoit74**
Fix #352 

Changes in logic around finding the ZIM illustration:
- build a sorted list of potential icons to use (instead of just a "random" one)
- prefer to use best icon in this list, either from WARC (preferably, if available) or from download (if available)
- if both fails for best icon, process next icon in the list (and so forth, until we find one working)
- if all fails, fallback to default scraperlib illustration

Searching inside the WARC is still preferred to download because it is expected to be quite fast: we already know the list of expected items and the icon is usually at the very beginning of the WARC (when present).

Note that the fact that this change contradicts significantly with what has been discussed and decided in https://github.com/openzim/warc2zim/issues/202, since there is a significant chance we will download the illustration.

In #202, we said that there is probably no situation where the best icon is not already present inside the WARC and should be downloaded. This is wrong. 

This change is grounded on a real use case: https://womenshistory.si.edu/. In this use case, we have only two WARC items fetched by the crawler:
- https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon.ico (which is 16x16)
- https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/favicon-32x32.png (which is 32x32)

Both are too small for a ZIM illustration and will need to be upscaled. This is not appropriate because scraper could know that best icon possible is https://womenshistory.si.edu//sites/default/themes/si_sawhm/favicons/android-chrome-192x192.png, and this file is available for download. This is now what will happen with the change in this PR.

### CODE REVIEW

File: src/warc2zim/converter.py [hunk 1]
```diff
@@ -648,43 +638,57 @@ def retrieve_illustration(self):
...
+                    path = normalize(HttpUrl(url))
+                    if path != favicon_path:
+                        continue
                     if (
                         record.http_headers
                         and record.http_headers.get_statuscode() != "200"
                     ):  # pragma: no cover
-                        logger.warning("WARC record for favicon is unusable")
+                        logger.warning(
+                            f"WARC record for favicon {favicon_path.value} is unusable"
```
Comment (rgaudin) [inline src/warc2zim/converter.py:657]: Would be the case for redirects right?

File: src/warc2zim/converter.py [hunk 2]
```diff
@@ -648,43 +638,57 @@ def retrieve_illustration(self):
...
-            # download favicon_url (might be custom URL, not present in WARC records)
-            if not self.illustration and self.favicon_url:
+            else:
+                # try to download icon as illustration
                 try:
+                    logger.debug(f"Downloading favicon at {favicon_url.value}")
                     dst = io.BytesIO()
-                    if not stream_file(self.favicon_url, byte_stream=dst)[0]:
+                    if not stream_file(favicon_url.value, byte_stream=dst)[0]:
                         raise OSError(
```
Comment (rgaudin) [inline src/warc2zim/converter.py:671]: Shouldn't we `continue` instead?

File: src/warc2zim/icon_finder.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            y_dim = min(y.width, y.height)
+            if x_dim == ZIM_ILLUSTRATION_SIZE:
+                if y_dim != ZIM_ILLUSTRATION_SIZE:
+                    return 1  # x is perfect, y not
+                return 0  # both images are perfect
+            if x_dim > ZIM_ILLUSTRATION_SIZE:
+                if y_dim == ZIM_ILLUSTRATION_SIZE:
+                    return -1  # y is perfect, x not
+                elif x_dim < y_dim:
+                    return -1  # perfer biggest icon (y)
```
Comment (rgaudin) [inline src/warc2zim/icon_finder.py:40]:
```suggestion
                    return -1  # prefer biggest icon (y)
```

File: src/warc2zim/icon_finder.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            y_dim = min(y.width, y.height)
+            if y_dim >= ZIM_ILLUSTRATION_SIZE and y_dim >= ZIM_ILLUSTRATION_SIZE:
+                return -1  # prefer y which is known to be big enough
+            else:
+                return 0  # y has height but it is too small, x we don't know
+        else:  # all other cases where we miss some sizes in x and y
+            return 0
+
+
+def sort_icons(icons: Iterable[IconUrlAndSize]) -> list[IconUrlAndSize]:
```
Comment (rgaudin) [inline src/warc2zim/icon_finder.py:73]: This is misleading. It's not sorting the iterable, it's returning a sorted version. Given stdlib already has the sorted (return) and sort (in place) concepts, we should stick to it.

```suggestion
def get_sorted_icons(icons: Iterable[IconUrlAndSize]) -> list[IconUrlAndSize]:
```

**Review by rgaudin — CHANGES_REQUESTED** [review]
This is a useful feature but I don't think it's wise to trust html only. I understand it's easier/faster because we don't have to download anything (for the cases there are downloads) but I think it would be more robust if checking the actual images.
This would have the added benefits of excluding those that cant convert to PNG and also simplify code since we'd not have images without dimensions.

I'm also a bit troubled by how the provided favicon is treated. My first opinion was to consider that the user is always right. Though checking that it's appropriate is a good thing.
I guess what itches me is the fact that user can specify something and we'll use something else without it mentioned in usage. That's in the rare case where specified is not exactly right and there's a better favicon in WARC.

File: src/warc2zim/converter.py (hunk 1 above)
Comment (benoit74) [reply src/warc2zim/converter.py:657]: This is indeed a limitation of current code. Never seen in the wild, so probably not something to take care of for now. This code is more targeted at 404.

File: src/warc2zim/icon_finder.py (hunk 4 above)
Comment (benoit74) [reply src/warc2zim/icon_finder.py:73]: Agreed

**Comment by benoit74** [comment]
Probably replaced by https://github.com/openzim/warc2zim/pull/372, keeping it for now until we decide which approach we merge

**Pull request closed without merging**